### PR TITLE
73: Integrate benchmark util and add verify + benchmark test

### LINKED ISSUE #73: Add benchmarks and unit tests to all-sky imager
Add unit tests and some benchmarks (for varying N baselines and varying N pixels) to the Python all-sky imager. 
Take into account the pytest-bencher integration.

Currently, the numba implementation fails on jitting the function due to some type confusion: https://github.com/astron-rd/PACE/actions/runs/29008986279/job/86087567483#step:5:443

### PR DESCRIPTION
**Pull request by Dantali0n**
This integrates a verification and benchmark test for the Python all-sky imaging.

1. It relies on several test and verification files added to the tests directory. These files are currently npz, we can convert to hdf5 but it would be extra work.
2. The measurement subfolder in the tests directory contains classes that can be reused for other benchmarks.
3. Features of benchmark class include;
   1. Warmup
   2. Statistics, mean, min, max, std
   3. PMT energy integration, supports aggregating multiple backends

The test is pretty basic as it compares images against a 'known good'.

### CODE REVIEW

File: all-sky/python/tests/measurements/measure.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+class Measure:
+    def __init__(self, settings: Settings = None):
+        self.measures: list[MeasureData] = []
+
+        self.pmt = False
+        if PMT_AVAILABLE:
+            logger.info("Using PMT RAPL power monitoring")
+            self.pms = []
+            self.pms.append(pmt.create("rapl"))
```
Comment (Dantali0n) [inline all-sky/python/tests/measurements/measure.py]: Create logic to decide on backends here.

**Review by csbnw — COMMENTED** [review]
Altough the changes look good to me, I would like to run the code once before I approve it. Could you write a consise `README.md` explaining the environment setup and what commands to run?

**Comment by Dantali0n** [comment]
> Altough the changes look good to me, I would like to run the code once before I approve it. Could you write a consise `README.md` explaining the environment setup and what commands to run?

Those changes aren't added because the README.md with that content already exists. See:

https://github.com/astron-rd/PACE/blob/main/all-sky/python/README.md

File: all-sky/python/tests/test_numba.py
Comment (bhaaksema) [inline all-sky/python/tests/test_numba.py]: Perhaps remove this placeholder test?

File: all-sky/python/tests/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    return (
+        visibilities,
+        baselines,
+    )
+
+
+def measure_imager(fn: Callable, settings: Settings):
+
+    visibilities, baselines = _load_npy(settings)
```
Comment (bhaaksema) [inline all-sky/python/tests/benchmark.py]:
```suggestion
def measure_imager(fn: Callable, settings: Settings):
    visibilities, baselines = _load_npy(settings)
```

File: all-sky/python/tests/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    path, file = settings.visibilities_path.rsplit("/", 1)
+    visibilities = np.load(files(path).joinpath(file))
+
+    path, file = settings.baselines_path.rsplit("/", 1)
+    baselines = np.load(files(path).joinpath(file))
+
+    return (
+        visibilities,
+        baselines,
+    )
```
Comment (bhaaksema) [inline all-sky/python/tests/benchmark.py]:
```suggestion
    return (visibilities, baselines)
```

File: all-sky/python/tests/measurements/measure.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        if self.pmt:
+            self._run_pmt_rapl(fn)
+        else:
+            self._run_time(fn)
+
+    def warmup(self, fn: Callable[[int, int], None]):
+        """Do warmup runs"""
+
+        for _ in range(self.settings.warmup):
+            fn(self.settings.image_size_x, self.settings.image_size_y)
```
Comment (bhaaksema) [inline all-sky/python/tests/measurements/measure.py]: Are you using [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/usage.html)? If so, it already has a built-in warmup feature.

Comment (Dantali0n) [reply all-sky/python/tests/measurements/measure.py]: I am not because I assumed we wanted to move away from that, or is that more a maybe.

Comment (bhaaksema) [reply all-sky/python/tests/measurements/measure.py]: To be fair, this is a bit of an unknown still. Probably be best to leave this until we know more.

**Review by bhaaksema — APPROVED** [review]

**Comment by bhaaksema** [comment]
`all_sky_jax.py` sets `jax_exec_time_optimization_effort = 1.0` with the comment "Gives another ~60% speedup on CPU". JAX 0.11 removed that option in favour of `jax_optimization_level`, so the line was updated in commit 7c2ff75, and I wondered if the new option gives a similar speedup. Below are measurements of the JAX imager with no flag, with the original flag, and with its replacement, on JAX versions 0.6.2, 0.10.2 and the current 0.11.1, on an x86 Linux VM and an Apple-silicon Mac.

**Setup**

- Kernel: `sky_imager_jax_ravel_real_jit`, 256x256, one visibility; wall time per call with `block_until_ready`, mean over the timed calls (x86: 10 warm-up + 15 timed; macOS: 30 + 30).
- none: neither flag set (JAX defaults).
- Flag A: `jax.config.update("jax_exec_time_optimization_effort", 1.0)`, the original line; removed in JAX 0.11.
- Flag B: `jax.config.update("jax_optimization_level", "O3")`, its replacement; available since JAX 0.5.3.
- Linux x86: GitHub Codespace, Intel Xeon Platinum 8370C @ 2.8 GHz, 2 vCPU, 7 GiB, Python 3.13.
- macOS arm64: Apple M1, 8 cores (4P + 4E), 16 GiB, macOS 27.0, Python 3.13.

| jax | platform | none | Flag A | Flag B |
|---|---|---|---|---|
| 0.6.2 | Linux x86 | 2.71 s | 2.67 s | 2.74 s |
| 0.10.2 | Linux x86 | 7.26 s | 7.24 s | 7.38 s |
| 0.11.1 | Linux x86 | 2.26 s | n/a | 2.25 s |
| 0.6.2 | macOS arm64 | 0.568 s | 0.557 s | 0.560 s |
| 0.10.2 | macOS arm64 | 1.145 s | 1.150 s | 1.149 s |
| 0.11.1 | macOS arm64 | 0.390 s | n/a | 0.397 s |

Std: 0.04-0.23 s on x86, 0.01-0.02 s on macOS; every within-row difference is inside it. Flag A is n/a on 0.11.1 because that release removed the option. Neither flag has a measurable effect on this kernel; the claimed ~60% does not reproduce on any platform or JAX version.

The 0.10.2 rows are 2-3x slower for a reason unrelated to the flags: the XLA CPU compiler in that series generated slow code for the kernel's final `mean` over the baseline axes, which 0.11 fixed.

**Comment by Dantali0n** [comment]
Very interesting, it seems to do pretty much nothing at all in your scenarios and on your hardware. I propose to drop the statement altogether but still set the optimization level to the highest available.

**Comment by Dantali0n** [comment]
Thanks for the fixes was meaning to get to this for quite some time. Finally we can merge.
